staging/src/k8s.io/apiserver/plugin/pkg/kmsprovider/vault/vault.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package vault implements envelop encryption provider based on Vault KMS
+package vault
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+	"sync"
+
+	"github.com/ghodss/yaml"
+
+	"k8s.io/apiserver/pkg/storage/value/encrypt/envelope"
+)
+
+const vaultPrefix = "vault"
+
+//EnvelopeConfig contains connection information for Vault transformer
+type EnvelopeConfig struct {
+	// The names of encryption key for Vault transit communication
+	KeyNames []string `json:"key-names"`
+	// Vault listen address, for example https://localhost:8200
+	Address string `json:"addr"`
+
+	// Token authentication information
+	Token string `json:"token"`
+
+	// TLS certificate authentication information
+	ClientCert string `json:"client-cert"`
+	ClientKey  string `json:"client-key"`
+
+	// AppRole authentication information
+	RoleID   string `json:"role-id"`
+	SecretID string `json:"secret-id"`
+
+	// CACert is the path to a PEM-encoded CA cert file to use to verify the
+	// Vault server SSL certificate.
+	CACert string `json:"ca-cert"`
```

Comment:
We use camelCase `roleId`, `secretId`.

Also, how `vaultCACert`?

Reply:
I changed CACert to vaultCACert for code readability.
The fields need to be exportable (yaml.unmarshall uses the interface), hence they start with Capital letters. I am following general two letter Acronym practice.

Reply:
> I changed CACert to vaultCACert for code readability.
> The fields need to be exportable (yaml.unmarshall uses the interface), hence they start with Capital letters. I am following general two letter Acronym practice.

For the value of the `json` tag.  Unless this must match an external format, you should be consistent with the project

```go
CACert string `json:"caCert"`
```

Reply:
done, for all json tags